pkg/targets/proxy.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package targets
+
+import (
+	"context"
+	"io"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"github.com/thanos-io/thanos/pkg/targets/targetspb"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Proxy implements targetspb.Targets gRPC that fanouts requests to given targetspb.Targets and deduplication on the way.
+type Proxy struct {
+	logger  log.Logger
+	targets func() []targetspb.TargetsClient
+}
+
+func RegisterTargetsServer(targetsSrv targetspb.TargetsServer) func(*grpc.Server) {
+	return func(s *grpc.Server) {
+		targetspb.RegisterTargetsServer(s, targetsSrv)
+	}
+}
+
+// NewProxy returns new targets.Proxy.
+func NewProxy(logger log.Logger, targets func() []targetspb.TargetsClient) *Proxy {
+	return &Proxy{
+		logger:  logger,
+		targets: targets,
+	}
+}
+
+func (s *Proxy) Targets(req *targetspb.TargetsRequest, srv targetspb.Targets_TargetsServer) error {
+	var (
+		g, gctx  = errgroup.WithContext(srv.Context())
+		respChan = make(chan *targetspb.TargetDiscovery, 10)
+		targets  []*targetspb.TargetDiscovery
+	)
+
+	for _, targetsClient := range s.targets() {
+		rs := &targetsStream{
+			client:  targetsClient,
+			request: req,
+			channel: respChan,
+			server:  srv,
+		}
+		g.Go(func() error { return rs.receive(gctx) })
+	}
+
+	go func() {
+		_ = g.Wait()
+		close(respChan)
+	}()
+
+	for resp := range respChan {
+		targets = append(targets, resp)
+	}
+
+	if err := g.Wait(); err != nil {
+		level.Error(s.logger).Log("err", err)
+		return err
+	}
+
+	for _, t := range targets {
+		if err := srv.Send(targetspb.NewTargetsResponse(t)); err != nil {
+			return status.Error(codes.Unknown, errors.Wrap(err, "send targets response").Error())
+		}
+	}
+
+	return nil
+}
+
+type targetsStream struct {
+	client  targetspb.TargetsClient
+	request *targetspb.TargetsRequest
+	channel chan<- *targetspb.TargetDiscovery
+	server  targetspb.Targets_TargetsServer
+}
+
+func (stream *targetsStream) receive(ctx context.Context) error {
```

Comment:
I wonder if we can reuse anything from StoreAPI proxy code :thinking: 

But otherwise it looks valid (:

Reply:
@bwplotka 
I can't figure out which part of StoreAPI proxy code and how I can use here. :(

As I can see currently rules proxy uses it's own code for similar method.

Maybe we can skip this optimization, then implement discovery API and after that see what we can change in both shared- rules and targets implementation?

What do you think about it?